ui/raidboss/data/07-dt/ultimate/futures_rewritten.ts

```diff
@@ -30,36 +67,206 @@ const p2KnockbackDirs: DirectionOutput8[] = [
+    shortFire: [] as string[],
+    mediumFire: '',
+    longFire: '',
+  },
+  ice: '',
+});
+
+// Helper for Ultimate Relativity that finds relative North based on the yellow-tethered lights
+// It takes an array of dir nums (e.g. 0-8), finds the two dir nums that have a single gap
+// between them (e.g. 1 and 3) -- the apex of the "Y" -- and returns the dir num of the gap.
+const findURNorthDirNum = (dirs: number[]): number => {
+  for (let i = 0; i < dirs.length; i++) {
+    for (let j = i + 1; j < dirs.length; j++) {
+      const [dir1, dir2] = [dirs[i], dirs[j]];
+      if (dir1 === undefined || dir2 === undefined)
+        return -1;
+      const diff = Math.abs(dir1 - dir2);
+      if (diff === 2)
+        return Math.min(dir1, dir2) + 1;
+      else if (diff === 6) // wrap around
+        return (Math.max(dir1, dir2) + 1) % 8;
+    }
+  }
+  return -1;
+};
+
+type ApocDebuffLength = 'short' | 'medium' | 'long' | 'none';
+type ApocDebuffMap = Record<ApocDebuffLength, string[]>;
+
+const p3UROutputStrings = {
+  yNorthStrat: {
+    en: '${debuff} (${dir})',
+  },
+  dirCombo: {
+    en: '${inOut} + ${dir}',
+  },
+  fireSpread: {
+    en: 'Fire - Spread',
+  },
+  dropRewind: {
+    en: 'Drop Rewind',
+  },
+  baitStoplight: {
+    en: 'Bait Stoplight',
+  },
+  avoidStoplights: {
+    en: 'Avoid stoplights',
+  },
+  stack: Outputs.stackMarker,
+  middle: Outputs.middle,
+  out: Outputs.out,
+};
+
 export interface Data extends RaidbossData {
+  readonly triggerSetConfig: {
+    sinboundRotate: 'aacc' | 'addposonly'; // aacc = always away, cursed clockwise
+    ultimateRel: 'yNorthDPSEast' | 'none';
+    apoc: 'dpsNE-CW' | 'none';
+  };
+  // General
+  phase: Phase | 'unknown';
   actorSetPosTracker: { [id: string]: NetMatches['ActorSetPos'] };
+  // P1 -- Fatebreaker
   p1ConcealSafeDirs: DirectionOutput8[];
   p1StackSpread?: 'stack' | 'spread';
   p1FallOfFaithTethers: ('fire' | 'lightning')[];
+  // P2 -- Usurper Of Frost
   p2QuadrupleFirstTarget: string;
   p2QuadrupleDebuffApplied: boolean;
-  p2IcicleImpactStart: DirectionOutput8;
+  p2IcicleImpactStart: DirectionOutput8[];
   p2AxeScytheSafe?: 'in' | 'out';
   p2FrigidStoneTargets: string[];
+  p2KBShivaDir?: DirectionOutput8;
+  p2LightsteepedCount: number;
+  p2LightRampantPuddles: string[];
+  p2SeenFirstHolyLight: boolean;
+  // P3 -- Oracle of Darkness
+  p3RelativityRoleCount: number;
+  p3RelativityDebuff?: RelativityDebuff;
+  p3RelativityRoleMap: ReturnType<typeof newRoleMap>;
+  p3RelativityStoplights: { [id: string]: NetMatches['AddedCombatant'] };
+  p3RelativityYellowDirNums: number[];
+  p3RelativityMyDirStr: string;
+  p3ApocDebuffCount: number;
+  p3ApocDebuffs: ApocDebuffMap;
+  p3MyApocDebuff?: ApocDebuffLength;
+  p3ApocFirstDirNum?: number;
+  p3ApocRotationDir?: 1 | -1; // 1 = clockwise, -1 = counterclockwise
+  p3CalledApoc: boolean;
 }
 
 const triggerSet: TriggerSet<Data> = {
   id: 'FuturesRewrittenUltimate',
   zoneId: ZoneId.FuturesRewrittenUltimate,
```

Comment:
```suggestion
  zoneId: ZoneId.FuturesRewrittenUltimate,
  comments: {
    en: 'Triggers: P1-3 / Timeline: P1-5',
  },
```
Adds the above comment to the coverage report.